Engaging Crowds: Subject set completeness

eatyourgreens opened the pull request:
Adds a query which gets available subject numbers, for a given workflow, from the Cellect API. 
Removes included subject sets from the workflows API query (see #678.) Subject sets are fetched separately, by ID, from the Panoptes API. This increases the number of API requests required to build each page, but allows us to limit subject set requests to grouped workflows. 
Updates the API mocks, in data-fetching tests, to include grouped/non-grouped cases and to add the `/subject_sets` endpoint.
Wraps the data-fetching requests in try/catch, so that we can log API errors to Sentry.
![HMS NHS home page, showing one subject set as 13% complete, and the other as 36% complete, for the first workflow.](https://user-images.githubusercontent.com/59547/108056476-36d84a00-7049-11eb-9313-3c4298e42f3f.png)

Package:
app-project

Closes #1956.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

beckyrother reviewed (changes requested):
Looks great! Just looks like there's a bit too much space between the last line and the bottom of the box – should just be 30px (the same as on the sides, which look right). Thanks!

![108056476-36d84a00-7049-11eb-9313-3c4298e42f3f_png__2462×1388_](https://user-images.githubusercontent.com/1392130/108089102-d6440f80-703e-11eb-8feb-9e6738abb2cb.png)

eatyourgreens commented:
@beckyrother Should be fixed now. This screenshot's from Firefox, but should be the same in other browsers.
![Subject set cards in Firefox, with smaller padding at the bottom of each card.](https://user-images.githubusercontent.com/59547/108376123-8ca02400-71fa-11eb-8714-3c37446525e4.png)

beckyrother reviewed (approved):
Looks great! Thanks Jim :)

shaunanoordin reviewed (changes requested):
## PR Review

This PR adds a "subject set completeness" reading to the Subject Set Selection menu (where applicable).

- The "subject set completeness" readout is a basic new UI element
- The code connects to Cellect to get the completeness stats (see Dev Notes below) ❕ 

Testing looks good on HMS NHS, but I ran into WTFs in other scenarios.

### Dev Notes

**The question of Cellect** ⚠️ ❔ 

Observation:
- this PR adds a fetch to the Cellect service (cellect.zooniverse.org) to fetch stats for Subject Sets.
- This fetch sub-action occurs on the main "classify" action logic route. i.e. it happens as soon as the user fetches the page, and isn't optional.

Thoughts:
- Concern: cellect.zooniverse.org seems to be a supplementary external service (i.e. not core to the classifier's main purpose) but it's tied to the main logic route.
- Opinion: If Cellect has issues for any reason, we _shouldn't_ allow it to either crash or slow down the main classifier experience for the user.

Recommendations(?):
- Pop a try-catch on the fetch-from-Cellect action...
- ...and drop the fetch timeout to, like, 5 seconds? If possible?

Level of concern, from 1 () to 10 (somebody just ran `rm -fr *` on the server root): 6? 7?

### Testing

Tested with macOS10+Chrome88

**Scenario 1:** HMS NHS (production)

This is the primary target project, and has 2 workflows with 2 subject sets each.

URLs:
- http://localhost:3000/projects/msalmon/hms-nhs-the-nautical-health-service?env=production
- http://localhost:3000/projects/msalmon/hms-nhs-the-nautical-health-service/classify?env=production

Testing Steps:
 - Go to the project page and click Classify, OR go to the root Classify URL directly.
 - You'll see the Workflow Selection screen. Pick any workflow.
 - You'll see the Subject Set Selection screen.
 - Observe that each Subject Sets "card" has an additional "50% complete" etc readout

This part LGTM! 👍

However... 

 **Scenario 2:** a bunch of other projects (staging and production)

URLs:
- http://localhost:3000/projects/darkeshard/transformers/classify?env=staging
- http://localhost:3000/projects/brooke/i-fancy-cats?env=staging
- http://localhost:3000/projects/darkeshard/prototype-2015?env=production
- http://localhost:3000/projects/mschwamb/planet-four?env=production

Results:
- Unable to view any of those pages!
- Website fully crashes with the error message:
  ```
  Server Error
  Error: Error serializing `.workflows[0]` returned from `getServerSideProps` in "/projects/[owner]/[project]".
  Reason: `undefined` cannot be serialized as JSON. Please use `null` or omit this value all together.
  ```

<img width="713" alt="Screen Shot 2021-02-19 at 00 08 50" src="https://user-images.githubusercontent.com/13952701/108438216-b67f3800-7246-11eb-8548-2adfe8680102.png">

🤷‍♂️ ❓ ❗ 

**Scenario 3:** more other projects

URLs:
- http://localhost:3000/projects/nora-dot-eisner/planet-hunters-tess?env=production

This part... looks good??

### Status

I'm popping a CR for this review:

1. The question of Cellect is something I'd like answered. If you think Cellect ain't something big to be worried about, I won't be too hard to convince.
2. The bugs with testing scenario 2 is something that I'm still trying to figure out. Why are only specific projects borking?? I chose a mix of projects with multiple WFs and single WFs.

Current status is 🤷 ❓ ❗

eatyourgreens commented on packages/app-project/src/helpers/fetchWorkflowsHelper/fetchWorkflowsHelper.js:
```diff
@@ -28,6 +29,14 @@ function fetchDisplayNames (language, activeWorkflows, env) {
     .then(createDisplayNamesMap)
 }
 
+async function fetchWorkflowCellectStatus(workflow) {
+  const workflowURL = `https://cellect.zooniverse.org/workflows/${workflow.id}/status`
+  const response = await fetch(workflowURL)
+  const body = await response.json()
+  const { groups } = body ?? {}
+  return groups
+}
```
Good point. If `getServerSideProps`, or `getStaticProps` errors, then the server will send back a 500 response. The same thing will happen if any of the Panoptes requests fail.

Ideally, I think we'd like to use `getStaticProps` and use static caching to reduce our dependency on API requests. I believe `getServerSideProps` runs on _every page transition_ so we are hitting the API a lot at the moment.

eatyourgreens commented on packages/app-project/src/helpers/fetchWorkflowsHelper/fetchWorkflowsHelper.js:
```diff
@@ -40,28 +49,39 @@ async function fetchPreviewImage (subjectSet, env) {
+  const subjectSetCounts = await fetchWorkflowCellectStatus(workflow)
+  const workflowSubjectSets = workflow.links.subject_sets
+    .map(subjectSetID => {
+      const subjectSet = subjectSets.find(subjectSet => subjectSet.id === subjectSetID)
+      subjectSet.availableSubjects = subjectSetCounts[subjectSetID]
```
Subject sets sent back from `fetchWorkflowData` are all subject sets from all workflows, so should include every linked set for each individual workflow.

It would be useful to test this for a realistic data set. The test project only has two workflows and two sets. What happens when you want to link 100 volumes of transcription material to your workflow?

eatyourgreens replied:
It turns out this is undefined for AnnoTate (but then, so are a bunch of other values that we would expect to exist, like workflow display name.)
http://local.zooniverse.org:3000/projects/drrogg/annotate?env=production

I suspect what's going on for AnnoTate is that completed sets aren't present in the Cellect response.

I'm now playing with the code to see what needs to be hacked to make AnnoTate work.
![Screenshot of the subject picker showing completed subject sets for AnnoTate.](https://user-images.githubusercontent.com/59547/108538465-baf33180-72d6-11eb-8f80-faf8d249194a.png)

eatyourgreens replied:
AnnoTate's workflow has 1198 linked subject sets, returned in pages of 20. I've completely rewritten the data fetching code to start dealing with that. There's a separate API request for the subject set data, using the subject IDs that are returned by Cellect.

eatyourgreens commented on packages/app-project/src/shared/components/SubjectSetPicker/components/SubjectSetCard/SubjectSetCard.js:
```diff
@@ -6,13 +6,15 @@ import { array, number, string } from 'prop-types'
   const placeholderUrl = `${assetPrefix}/subject-placeholder.png`
   const subjectURLs = subject ? subject.locations.map(location => Object.values(location)[0]) : []
   const alt = subject ? `Subject ${subject.id}` : 'Loading'
+  const completeness = 1 - (availableSubjects / set_member_subjects_count)
```
Good point. Does Panoptes let you create sets with no subjects in them?

shaunanoordin replied:
It's definitely possible to create Workflow with an empty Subject Set. I made a quick test, and the result is that the Classifier just shrugs and says "I guess there's no subject to load" (with a bunch of errors in the dev console) and displays a blank Task Area and blank Subject Viewer.

If you're curious: http://localhost:3000/projects/darkeshard/transformers/classify/workflow/3457

This is still in the "minor worries" category though.

eatyourgreens commented:
The errors from standard projects should be fixed now. I'd forgotten to check for `workflow.grouped` before making the request to Cellect. I've also beefed up the tests to include responses for grouped and non-grouped workflows in the mock API.

eatyourgreens commented:
If Cellect is down, I guess we'll lose sequential subject selection for Engaging Crowds. I think the fallback then might be the default Panoptes selector. Random selection maybe? So Cellect being unavailable might not just break the API requests here, it might break the subject queue too.

shaunanoordin commented:
## PR Review (Update)

First off, a big 👍 for making sure there's some branching logic that checks when `(!workflow.grouped)` - this certainly goes a long way in addressing unnecessary calls to Cellect.

Second, I ran test through the URLs and...

- ✅ http://localhost:3000/projects/msalmon/hms-nhs-the-nautical-health-service?env=production
- ✅ http://localhost:3000/projects/msalmon/hms-nhs-the-nautical-health-service/classify?env=production
- ✅ http://localhost:3000/projects/darkeshard/transformers/classify?env=staging
- ✅ http://localhost:3000/projects/brooke/i-fancy-cats?env=staging
- ❌  http://localhost:3000/projects/darkeshard/prototype-2015?env=production
- ✅ http://localhost:3000/projects/mschwamb/planet-four?env=production
- ✅ http://localhost:3000/projects/nora-dot-eisner/planet-hunters-tess?env=production

So far I've traced the issue of Prototype 2015 to the following:

- It's production Project 1292
- It has one active workflow, WF900
- For some reason, `fetchWorkflowsHelper` returns 900's displayName as `undefined`
- I'm still trying to figure out why

I'm filing this as a Monday problem. Yoinks!

shaunanoordin commented:
Here's what I found out:

Pinging `https://panoptes.zooniverse.org/api/translations/?translated_id=7792&translated_type=workflow` results in...

```
{
    "translations": [
        {
            "id": "6709",
            "translated_id": 7792,
            "translated_type": "Workflow",
            "language": "en",
            "strings": {
                "display_name": "Alternate identification",
                "tasks.T0.help": "",
                "tasks.T0.question": "Good or Bad?",
                "tasks.T0.answers.0.label": "Good",
                "tasks.T0.answers.1.label": "Bad",
                "tasks.T0.answers.2.label": "Morality is subjective and you cannot objectively judge robots based solely on your own moral viewpoint "
            },
            "string_versions": {
                "display_name": 6403569,
                "tasks.T0.help": 6403569,
                "tasks.T0.question": 6403569,
                "tasks.T0.answers.0.label": 6403569,
                "tasks.T0.answers.1.label": 6403569,
                "tasks.T0.answers.2.label": 6403569
            },
            "href": "/translations/6709",
            "created_at": "2018-09-26T12:12:31.622Z",
            "updated_at": "2019-02-22T11:39:51.587Z",
            "links": {
                "published_version": null,
                "workflow": "7792"
            }
        }
    ],
    "links": { ... }
}
```

BUT compare that to `https://panoptes.zooniverse.org/api/translations/?translated_id=900&translated_type=workflow` which results in...

```
{
    "translations": [],
    "links": { ... }
}
```

So Workflow 7792 has translations, but WF 900 doesn't? (They're both WFs in the same Project 1292)

I think the key takeaway here is that we can't 100% rely on /translations to always return resources??

eatyourgreens commented:
When was workflow 900 last updated? Anything updated in the last two years or so should have translations, I think.

eatyourgreens commented:
I've added `workflow.display_name` as a fallback for the workflow name, but I'm not sure we should do a lot of work to accommodate workflows that are more than 2 or 3 years old. Those workflows would error with our deployed production code.

eatyourgreens commented on packages/app-project/src/helpers/fetchWorkflowsHelper/fetchWorkflowsHelper.js:
```diff
@@ -28,6 +43,21 @@ function fetchDisplayNames (language, activeWorkflows, env) {
+      const body = await response.json()
+      groups = body.groups ?? {}
+    } catch (error) {
+      console.error(error)
```
Maybe send these errors to Sentry?

eatyourgreens commented on packages/app-project/src/helpers/fetchWorkflowsHelper/fetchWorkflowsHelper.js:
```diff
@@ -1,18 +1,32 @@
+    subject_sets = response.body.subject_sets
+    await Promise.allSettled(subject_sets.map(subjectSet => fetchPreviewImage(subjectSet, env)))
+  } catch (error) {
+    console.error(error)
```
Maybe send these errors to Sentry too?

eatyourgreens commented:
The tests expect `fetchWorkflowsHelper` to pass errors on to consumers, so I'm re-throwing caught errors for workflow requests.

srallen reviewed (commented):
This is just a conceptual question out of scope of this PR (which I'm happy to make a discussion issue for), but it seems like theres _a lot_ async code going on in this helper file, much more than its name `fetchWorkflowsHelper` would imply and I'm not just not clear why there isn't a workflow store for all this.

eatyourgreens commented:
I think that's historical, going back to #1077 when `fetchWorkflowsHelper` was added to fetch data for the project Hero component. It is doing a lot more work now, across more than one component. We could definitely add the workflow data to the `initialState` page prop, then hydrate a workflows store in the browser. That said, I think workflow state is static: once we've fetched workflows and subject sets from the API, they're never updated in the browser.

There's [an ongoing discussion to add static props to `_app.js`](https://github.com/vercel/next.js/discussions/10949), for cases like this. It doesn't make a lot of sense to be constantly fetching all this data from Panoptes when we could cache it across requests. Thanks to that discussion, I did discover [`next-plugin-preval`](https://github.com/ricokahler/next-plugin-preval#usage), which might be useful here.

eatyourgreens commented:
I've updated this to include the latest changes to our Sentry setup. I've also split the Node data-fetching code into `helpers/fetchWorkflowsHelper` and `helpers/fetchSubjectSets`.

shaunanoordin commented:
Following up from my previous [PR review,](https://github.com/zooniverse/front-end-monorepo/pull/2016#pullrequestreview-593719544) the functionality checks now all look good! I'm going to give the code a quick read to see what updates have happened since I last went through this. brb!

shaunanoordin reviewed (approved):
## PR Review (Update)

Follows [earlier PR review](https://github.com/zooniverse/front-end-monorepo/pull/2016#pullrequestreview-593719544)

Functionality tests look good, and code read looks good!

### Testing

Same tests as before, and everything's passing OK now.

**Scenario 1:** HMS NHS (production)
- http://localhost:3000/projects/msalmon/hms-nhs-the-nautical-health-service?env=production
- http://localhost:3000/projects/msalmon/hms-nhs-the-nautical-health-service/classify?env=production

Results: 🆗  Completeness value for each WF and SS are displaying properly on their selection screens.

**Scenario 2:** a bunch of other projects (staging and production)
- http://localhost:3000/projects/darkeshard/transformers/classify?env=staging
- http://localhost:3000/projects/brooke/i-fancy-cats?env=staging
- http://localhost:3000/projects/darkeshard/prototype-2015?env=production
- http://localhost:3000/projects/mschwamb/planet-four?env=production

Results: 🆗 These workflows load normally. I take it the "can't find translations" issue has been resolved, by using the `workflow.display_name` as a fallback? (i.e.: `const displayName = displayNames[workflow.id] || workflow.display_name `) Lookin' good!

**Scenario 3:** more other projects
- http://localhost:3000/projects/nora-dot-eisner/planet-hunters-tess?env=production

Results: 🆗

### Status

LGTM 👍 I genuinely can't find anything of issue with this PR in its current state. Let's gooo!

eatyourgreens commented on packages/app-project/src/helpers/fetchSubjectSets/fetchSubjectSets.js:
```diff
@@ -0,0 +1,65 @@
+  }
+}
+
+export default async function workflowSubjectSets(workflow, env) {
```
```suggestion
export default async function fetchSubjectSets(workflow, env) {
```
